clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -268,6 +280,13 @@ public void poll(long timeout) throws IOException {
         currentTimeNanos = endSelect;
         this.sensors.selectTime.record(endSelect - startSelect, time.milliseconds());
 
+        // Perform setup for instantly connected channels since they won't be included in selectedKeys()
+        if (connectableChannels.size() > 0)
+            for (KafkaChannel channel : connectableChannels) {
+                sensors.maybeRegisterConnectionMetrics(channel.id());
```

Comment:
I can trip this every time when running load tests. Our code starts 7 modules which start 64 consumers each which subscribe to 1 topic (unique per module) with 64 partitions. So, that's 448 consumers starting nearly simultaneously on a vmware VM with 8 cores. Out of 448 only 1-4 connections connect immediately. My guess is that those threads are yielding long enough for the connection to complete before the SocketChannel.connect() method does. It may be possible to trigger this by launching a few hundred threads to tie up the CPUs.

We unit test such things by having the the class under test accept a java.nio.channels.spi.SelectorProvider so we can provide mock instances of Selector and SocketChannel. I strongly recommend doing that as you will significantly improve the testability of the code.

I will run the latest patch tomorrow (Sat 3/19) when I'm back on-line and report back.